code/part-two/client/source/services/transactions.js

```diff
@@ -0,0 +1,166 @@
+// Functions for creating private keys and submitting transactions
+'use strict';
+
+import axios from 'axios';
+import { createHash } from 'crypto';
+import secp256k1 from 'sawtooth-sdk/signing/secp256k1';
+import {
+  Transaction,
+  TransactionHeader,
+  Batch,
+  BatchHeader,
+  BatchList
+} from 'sawtooth-sdk/protobuf';
+
+
+const hash = str => createHash('sha512').update(str).digest('hex');
+
+const FAMILY_NAME = 'cryptomoji';
+const FAMILY_VERSION = '1.0';
+const NAMESPACE = hash(FAMILY_NAME).slice(0, 6);
+const CONTEXT = new secp256k1.Secp256k1Context();
+
+// Returns a random 1-12 character string
+const getNonce = () => (Math.random() * 10 ** 18).toString(36);
+
+// Recursively fetches all of the keys in nested objects
+const listKeys = obj => {
+  if (!obj || typeof obj !== 'object') {
+    return [];
+  }
+
+  if (Array.isArray(obj)) {
+    return obj.reduce((keys, item) => {
+      return keys.concat(listKeys(item));
+    }, []);
+  }
+
+  const topKeys = Object.keys(obj);
+  const nestedKeys = topKeys.reduce((keys, key) => {
+    return keys.concat(listKeys(obj[key]));
+  }, []);
+  return topKeys.concat(nestedKeys);
+};
+
+// Encode payload as a sorted JSON buffer
+const encodePayload = payload => {
+  const sortedKeys = listKeys(payload).sort();
+  return Buffer.from(JSON.stringify(payload, sortedKeys));
+};
+
+// Returns a function to create new Transactions
+const getTxnCreator = (privateWrapper, publicKey) => encodedPayload => {
+  const header = TransactionHeader.encode({
+    signerPublicKey: publicKey,
+    batcherPublicKey: publicKey,
+    familyName: FAMILY_NAME,
+    familyVersion: FAMILY_VERSION,
+    inputs: [NAMESPACE],
+    outputs: [NAMESPACE],
+    nonce: getNonce(),
+    payloadSha512: hash(encodedPayload)
+  }).finish();
+
+  return Transaction.create({
+    header,
+    payload: encodedPayload,
+    headerSignature: CONTEXT.sign(header, privateWrapper)
+  });
+};
+
+// Returns a function to create batches
+const getBatchCreator = (privateWrapper, publicKey) => txns => {
+  const header = BatchHeader.encode({
+    signerPublicKey: publicKey,
+    transactionIds: txns.map(txn => txn.headerSignature)
+  }).finish();
+
+  return Batch.create({
+    header,
+    headerSignature: CONTEXT.sign(header, privateWrapper),
+    transactions: txns
+  });
+};
+
+// Encode a Batch in a BatchList
+const encodeBatch = batch => {
+  return BatchList.encode({ batches: [batch] }).finish();
+};
+
+// Submits an encoded BatchList to the API
+const submitBatchList = (batchList, shouldWait) => {
+  return axios({
+    method: 'POST',
+    url: 'api/batches',
+    data: batchList,
+    headers: { 'Content-Type': 'application/octet-stream' }
+  })
+    .then(({ data }) => {
+      if (!shouldWait) {
+        return true;
+      }
+
+      return axios.get(data.link + '&wait')
```

Comment:
minor suggestion: end the `.then(...)` block from ln 98 right after ln 103, and then attach the `.then(...) on ln 104 afterwards